drivers/input/input_gpio_keys.c

```diff
@@ -27,33 +28,41 @@ struct gpio_keys_pin_config {
 	/** Zephyr code from devicetree */
 	uint32_t zephyr_code;
 };
+
+struct gpio_keys_pin_data {
+	const struct device *dev;
+	struct gpio_keys_callback cb_data;
+	struct k_work_delayable work;
+	int8_t pin_state;
+};
+
 struct gpio_keys_config {
 	/** Debounce interval in milliseconds from devicetree */
 	uint32_t debounce_interval_ms;
 	const int num_keys;
 	const struct gpio_keys_pin_config *pin_cfg;
+	struct gpio_keys_pin_data *pin_data;
+	k_work_handler_t handler;
+	bool no_interrupts;
 };
 
-struct gpio_keys_pin_data {
-	const struct device *dev;
-	struct gpio_keys_callback cb_data;
-	struct k_work_delayable work;
-	int8_t pin_state;
+struct gpio_keys_data {
+#ifdef CONFIG_PM_DEVICE
+	atomic_t suspended;
+#endif
 };
 
 /**
  * Handle debounced gpio pin state.
  */
-static void gpio_keys_change_deferred(struct k_work *work)
+static void gpio_keys_poll_pin(const struct device *dev, int key_index)
 {
-	struct k_work_delayable *dwork = k_work_delayable_from_work(work);
-	struct gpio_keys_pin_data *pin_data = CONTAINER_OF(dwork, struct gpio_keys_pin_data, work);
-	const struct device *dev = pin_data->dev;
-	int key_index = pin_data - (struct gpio_keys_pin_data *)dev->data;
 	const struct gpio_keys_config *cfg = dev->config;
 	const struct gpio_keys_pin_config *pin_cfg = &cfg->pin_cfg[key_index];
+	struct gpio_keys_pin_data *pin_data = &cfg->pin_data[key_index];
+	int new_pressed;
 
-	const int new_pressed = gpio_pin_get(pin_cfg->spec.port, pin_cfg->spec.pin);
+	new_pressed = gpio_pin_get(pin_cfg->spec.port, pin_cfg->spec.pin);
 
 	LOG_DBG("gpio_change_deferred %s pin_state=%d, new_pressed=%d, key_index=%d", dev->name,
```

Comment:
Nit: the prefix does not match the function anymore. Wonder if you can drop it since DBG will print the function name anyways.

Reply:
hey good point, tweaked this and another debug print